datafusion/sql/src/expr/mod.rs

```diff
@@ -215,11 +216,28 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                         _ => (expr, false),
                     }
                 }
+
                 let expr =
-                    self.sql_expr_to_logical_expr(*obj, schema, planner_context)?;
+                    self.sql_expr_to_logical_expr(*expr, schema, planner_context)?;
+
+                let indexes = match *subscript {
+                    Subscript::Index { index } => vec![index],
+                    Subscript::Slice { lower_bound, upper_bound } => {
+                        let mut indexes = vec![];
+                        if let Some(lower_bound) = lower_bound {
+                            indexes.push(lower_bound);
+                        }
+                        if let Some(upper_bound) = upper_bound {
+                            indexes.push(upper_bound);
+                        }
+                        indexes
+                    }
+                };
```

Comment:
Now we pass the compilation. But it still fails `array.slt`:

```
External error: query failed: DataFusion error: Error during planning: Error during planning: Coercion from [Int64, Int64] to the signature ArraySignature(ArrayAndIndex) failed. and No function matches the given name and argument types 'array_element(Int64, Int64)'. You might need to add explicit type casts.
	Candidate functions:
	array_element(array, index)
[SQL] select make_array(1, 2, 3)[1:2], make_array(1.0, 2.0, 3.0)[2:3], make_array('h', 'e', 'l', 'l', 'o')[2:4];
at test_files/array.slt:859

Error: Execution("1 failures")
error: test failed, to rerun pass `-p datafusion-sqllogictest --test sqllogictests`
```

Reply:
For `make_array(1, 2, 3)[1:2]`, previous AST:

```
ArrayIndex {
obj: ...,
indexes: [
    JsonAccess {
        left: Value(Number("1", false)),
        operator: Colon,
        right: Value(Number("2", false)),
    },
],
}
```

new AST:

```
Subscript {
expr: ...,
subscript: Slice {
    lower_bound: Some(Value(Number("1", false))),
    upper_bound: Some(Value(Number("2", false))),
},
}
```

Reply:
Seems we need adapt `plan_indices` with this new `Subscript` struct and it may be wrapped into a `SQLExpr` struct.

Reply:
Also, the patch still fails to support `select make_array(1, 2, 3)[1:2:2], make_array(1.0, 2.0, 3.0)[2:3:2], make_array('h', 'e', 'l', 'l', 'o')[2:4:2];`.